src/libraries/System.Data.Common/src/System/Data/Common/BooleanStorage.cs

```diff
@@ -12,7 +12,7 @@ internal sealed class BooleanStorage : DataStorage
     {
         private const bool defaultValue = false;
 
-        private bool[] _values;
+        private bool[] _values = default!; // Late-initialized
```

Comment:
What does the compiler complain about with this if you don't add the default assignment?

Reply:
If there's no assignment, the compiler complains that the constructor doesn't initialize it (all non-nullable fields/properties must be initialized)

Reply:
Ah. We can't explain lazy-init to the compiler so we have to either go with upfront assign and non-nullable or nullablle when the reality is just-in-time not nullable. The number of edges cases where nullability just doesn't fit quite right is annoying. I think what we really want here is the ability to ignore the field for the purposes of nullability checking but to ensure that all methods which can touch it do the check correctly.

Reply:
Yeah, this is an example of an annoying case. At the end of the day, the type is simply poorly-designed for good nullability handling, since there's an assumption that the storage (i.e. the bool array) is always set externally at some point after construction, but before the storage is actually used. Unfortunately doing any refactoring is definitely beyond our scope here.

At the end of the day there really are two choices:
1. Mark the field as nullable, requiring checks or null-forgiving (bang) operators almost everywhere where the field is accessed, or
2. Mark the field as non-nullable (what I chose to do here), with the implicit assumption that it is late-initialized.

In general, as this is a private field it doesn't matter that much, and we can always change this later if we choose without consequences.